users/luescher/setups/hybrid/gmm_system.py

```diff
@@ -0,0 +1,845 @@
+            "train_{}_alnt_bundle_last".format(name),
+            self.jobs[corpus]["train_{}".format(name)]
+            .selected_alignment_jobs[-1]
+            .alignment_bundle,
+        )
+        tk.register_output(
+            "train_{}_mix_last".format(name),
+            self.jobs[corpus]["train_{}".format(name)]
+            .selected_mixture_jobs[-1]
+            .mixtures,
+        )
+
+    # -------------------- recognition  --------------------
+
+    def recognition(
+        self,
+        name,
+        iters,
+        corpus,
+        feature_flow,
+        feature_scorer,
+        pronunciation_scale,
+        lm_scale,
+        search_params,
+        rtf,
+        mem,
+        parallelize_conversion,
+        lattice_to_ctm_kwargs,
+        **kwargs,
+    ):
+        with tk.block(f"{name}_recognition"):
+            optimize_am_lm_scale = self.gmm_args.monophone_recognition_args.pop(
+                "optimize_am_lm_scale", False
+            )
+            recog_func = self.recog_and_optimize if optimize_am_lm_scale else self.recog
+
+            for it in iters:
+                recog_func(
+                    name=f"{name}-{it:02d}",
+                    prefix=f"recog_{corpus}_{name}/",
+                    corpus=corpus,
+                    flow=feature_flow,
+                    feature_scorer=list(feature_scorer) + [it - 1],
+                    pronunciation_scale=pronunciation_scale,
+                    lm_scale=lm_scale,
+                    search_parameters=search_params,
+                    rtf=rtf,
+                    mem=mem,
+                    parallelize_conversion=parallelize_conversion,
+                    lattice_to_ctm_kwargs=lattice_to_ctm_kwargs,
+                    **kwargs,
+                )
+
+    # -------------------- run setup  --------------------
+
+    def run(self, steps=tuple("all")):
+        assert len(steps) > 0
+        if len(steps) == 1 and steps[0] == "all":
+            steps = ["extract", "mono", "cart", "tri", "vtln", "sat", "vtln+sat"]
+
+        if "init" in steps:
+            print(
+                "init needs to be run manually. provide: gmm_args, {train,dev,test}_inputs"
+            )
+            sys.exit(-1)
+
+        if "extract" in steps:
+            self.extract_features(feat_args=self.gmm_args.feature_extraction_args)
+
+        # ---------- Monophone ----------
+        if "mono" in steps:
+            for trn_c in self.train_corpora:
+                self.monophone_training(
+                    "mono",
+                    trn_c,
+                    self.gmm_args.linear_alignment_args,
+                    **self.gmm_args.monophone_training_args,
+                )
+
+                for dev_c in self.dev_corpora:
+                    pass
+
+                for tst_c in self.test_corpora:
+                    pass
+
+        # ---------- CaRT ----------
+        if "cart" in steps:
+            for c in self.train_corpora:
```

Comment:
Is estimating a separate CART for each training corpus the intended bahavior? I would discuss that we want the same cart for every corpus.

In fact, maybe we want to merge all (or a variable selection of all) training corpora for each training step.

Reply:
This is not what I wanted. If you have all corpora and audio files available you would merge all available corpora via MergeCorporaJob. Then you would set the merged single corpus file as a starting point. So you would actually only have one corpus within the class.
Later on, you might want to add corpora (for speed perturbation or TTS) and then redo for example the cart step. But this is not reflected in the "default' run function...